zkevm-circuits/src/bytecode_circuit/circuit.rs

```diff
@@ -0,0 +1,1041 @@
+                meta.query_advice(bytecode_table.is_code, Rotation::next()),
+                1.expr(),
+            );
+
+            cb.require_equal(
+                "next.hash == cur.hash",
+                meta.query_advice(bytecode_table.code_hash, Rotation::next()),
+                meta.query_advice(bytecode_table.code_hash, Rotation::cur()),
+            );
+
+            cb.require_equal(
+                "next.value_rlc == next.value",
+                meta.query_advice(value_rlc, Rotation::next()),
+                meta.query_advice(bytecode_table.value, Rotation::next()),
+            );
+
+            cb.gate(and::expr(vec![
+                meta.query_fixed(q_enable, Rotation::cur()),
+                not::expr(meta.query_fixed(q_last, Rotation::cur())),
+                is_header_to_byte(meta),
+            ]))
+        });
+
+        // When is_byte_to_byte ->
+        // assert next.length == cur.length
+        // assert next.index == cur.index + 1
+        // assert next.hash == cur.hash
+        // assert next.value_rlc == cur.value_rlc * randomness + next.value
+        // if cur.is_code:
+        //     assert next.push_data_left == cur.push_data_size
+        // else:
+        //     assert next.push_data_left == cur.push_data_left - 1
+        meta.create_gate("Byte to Byte row", |meta| {
+            let mut cb = BaseConstraintBuilder::default();
+
+            cb.require_equal(
+                "next.length == cur.length",
+                meta.query_advice(length, Rotation::next()),
+                meta.query_advice(length, Rotation::cur()),
+            );
+
+            cb.require_equal(
+                "next.index == cur.index + 1",
+                meta.query_advice(bytecode_table.index, Rotation::next()),
+                meta.query_advice(bytecode_table.index, Rotation::cur()) + 1.expr(),
+            );
+
+            cb.require_equal(
+                "next.hash == cur.hash",
+                meta.query_advice(bytecode_table.code_hash, Rotation::next()),
+                meta.query_advice(bytecode_table.code_hash, Rotation::cur()),
+            );
+
+            // TODO: check this
+            cb.require_equal(
+                "next.value_rlc == cur.value_rlc * randomness + next.value",
+                meta.query_advice(value_rlc, Rotation::next()),
+                meta.query_advice(value_rlc, Rotation::cur()) * challenges.keccak_input()
+                    + meta.query_advice(value, Rotation::next()),
+            );
+
+            cb.require_equal(
+                "next.push_data_left == cur.is_code ? cur.push_data_size : cur.push_data_left - 1",
+                meta.query_advice(push_data_left, Rotation::next()),
+                select::expr(
+                    meta.query_advice(bytecode_table.is_code, Rotation::cur()),
+                    meta.query_advice(push_data_size, Rotation::cur()),
+                    meta.query_advice(push_data_left, Rotation::cur()) - 1.expr(),
+                ),
```

Comment:
I think a comment could explain the optimisation in case that readability is poorer

Reply:
@han0110 actually when is_code is 1, we don't substract 1, and when is_code is 0 we do substract 1. hence, this would be equivalent to push_data_size - not(is_code)

Reply:
Ah you are right!

Reply:
actually, this does not work, because in one case is push_data_SIZE when in the other it is push_data_LEFT - 1

Reply:
Yes... Sorry I didn't check the formula carefully.